src/seals/base_envs.py

```diff
@@ -179,44 +238,84 @@ def __init__(
             raise ValueError(
                 "Malformed transition_matrix:\n"
                 f"transition_matrix.shape: {transition_matrix.shape}\n"
-                f"{n_states} != {n_next_states}",
+                f"{transition_matrix.shape[0]} != {transition_matrix.shape[2]}",
+            )
+
+        if reward_matrix.shape != transition_matrix.shape[: len(reward_matrix.shape)]:
+            raise ValueError(
+                "transition_matrix and reward_matrix are not compatible:\n"
+                f"transition_matrix.shape: {transition_matrix.shape}\n"
+                f"reward_matrix.shape: {reward_matrix.shape}",
+            )
+
+        if observation_matrix.shape[0] != transition_matrix.shape[0]:
+            raise ValueError(
+                "transition_matrix and observation_matrix are not compatible:\n"
+                f"transition_matrix.shape[0]: {transition_matrix.shape[0]}\n"
+                f"observation_matrix.shape[0]: {observation_matrix.shape[0]}",
             )
 
         if initial_state_dist is None:
-            initial_state_dist = util.one_hot_encoding(0, n_states)
+            initial_state_dist = util.one_hot_encoding(0, transition_matrix.shape[0])
         if initial_state_dist.ndim != 1:
             raise ValueError(
                 "initial_state_dist has multiple dimensions:\n"
                 f"{initial_state_dist.ndim} != 1",
             )
-        if initial_state_dist.shape[0] != n_states:
+        if initial_state_dist.shape[0] != transition_matrix.shape[0]:
             raise ValueError(
                 "transition_matrix and initial_state_dist are not compatible:\n"
-                f"n_states = {n_states}\n"
+                f"number of states = {transition_matrix.shape[0]}\n"
                 f"len(initial_state_dist) = {len(initial_state_dist)}",
             )
 
-        if reward_matrix.shape != transition_matrix.shape[: len(reward_matrix.shape)]:
-            raise ValueError(
-                "transition_matrix and reward_matrix are not compatible:\n"
-                f"transition_matrix.shape: {transition_matrix.shape}\n"
-                f"reward_matrix.shape: {reward_matrix.shape}",
-            )
-
         self.transition_matrix = transition_matrix
         self.reward_matrix = reward_matrix
+        self.observation_matrix = observation_matrix
         self._feature_matrix = None
         self.horizon = horizon
         self.initial_state_dist = initial_state_dist
 
         super().__init__(
-            state_space=spaces.Discrete(n_states),
-            action_space=spaces.Discrete(n_actions),
+            state_space=self._construct_state_space(self.state_dim),
+            action_space=self._construct_action_space(self.action_dim),
+            observation_space=self._construct_obs_space(self.obs_dim, self.obs_dtype),
+        )
+
+    @staticmethod
+    def _construct_state_space(n_states: int) -> gym.Space:
```

Comment:
Yes, primarily that is the underlying point. Changing what the state space is by a subclass would require re-writing all the validation we do in `BaseTabularModelPOMDP`, or accepting optional custom `state_space` etc from the user on the base class. I found that overriding a constructor was cleaner.

Even though I must admit that I didn't think very carefully about how general the type signature should be, especially since they are static methods. 

I guess it depends on how likely you think it is that someone will ever want to override this. I think that a very obvious possible use case is changing the bounds of the box observation space, or requiring a different geometric shape that puts other constraints like a hypersphere or a simplex. But this might be totally overkill/unnecessary, I don't know.

Reply:
Leaving it as-is seems reasonable, thanks for the clarification.

Reply:
Thinking about it more, using a different state space seems tricky, as the user would also need to override the `transition` and `reward` function and anything else that takes `state`. But no harm leaving it as-is. Overriding `_construct_obs_space` makes a lot more sense.